pkg/ingester-rf1/ingester.go

```diff
@@ -571,32 +571,43 @@ func (i *Ingester) loop() {
 	for {
 		select {
 		case <-flushTicker.C:
-			//i.logger.Log("msg", "starting periodic flush")
-			i.flushCtx.lock.Lock() // Stop new chunks being written while we swap destinations - we'll never unlock as this flushctx can no longer be used.
-			currentFlushCtx := i.flushCtx
-			// APIs become unblocked after resetting flushCtx
-			segmentWriter, err := wal.NewWalSegmentWriter()
-			if err != nil {
-				// TODO: handle this properly
-				panic(err)
-			}
-			i.flushCtx = &flushCtx{
-				lock:            &sync.RWMutex{},
-				flushDone:       make(chan struct{}),
-				newCtxAvailable: make(chan struct{}),
-				segmentWriter:   segmentWriter,
-			}
-			close(currentFlushCtx.newCtxAvailable) // Broadcast to all waiters that they can now fetch a new flushCtx. Small chance of a race but if they re-fetch the old one, they'll just check again immediately.
-			// Flush the finished context in the background & then notify watching API requests
-			// TODO: use multiple flush queues if required
-			i.flushQueues[0].Enqueue(currentFlushCtx)
-
+			i.doFlushTick()
 		case <-i.loopQuit:
 			return
 		}
 	}
 }
 
+func (i *Ingester) doFlushTick() {
+	i.flushCtx.lock.Lock()
+	defer i.flushCtx.lock.Unlock()
```

Comment:
This probably doesn't hurt but we actually don't want to unlock this - the context is only locked once when its no longer accepting writes.
I wonder if its possible that this goroutine can gets pre-empted and another tick can fire before we swap in a new context? In that case we'd block the timer thread completely waiting on the lock... Best to leave the unlock here for now in that case!

Reply:
I think you are right!